### ISSUE TITLE
Use of Setbitrate during connection

### ISSUE THREAD

**Issue by astaikos316** [microsoft/MixedReality-WebRTC#161]
Does anyone know if it is possible to call setbitrate() during a call to allow adjustments of bandwidth depending on network conditions? I have been doing some trials and getting unexpected results.  Initial one-way bandwidth usage is measured at about 4Mbps.  When trying to run setbitrate() with new min, current, max values, bandwidth usage drops to about 2 Mbps but was expecting it to be lower.

**Comment by djee-ms** [comment]
Sorry @astaikos316 I didn't reply earlier because I have not tried myself and I never dug into the Google implementation to see exactly if it can be applied dynamically. We are just exposing the same-named `SetBitrates()` from Google's implementation, and it doesn't seem to be documented whether this is supported.

For reference in Google's code:
- `SetBitrates()` calls `RtpBitrateConfigurator::UpdateWithClientPreferences()`
- `RtpBitrateConfigurator` updates its `bitrate_config_`
- The config seems to be read only by `RtpTransportControllerSend::OnNetworkRouteChanged()`
But it is unclear when that callback is fired, and from the name I doubt it is very often.

Note that `PeerConnection.SetBitrate()` doesn't take a current bitrate parameter. The start bitrate, which is used in particular for the video encoder startup, has a comment saying "start/current" but I cannot recall why because I would be surprised this works. My recollection was that the network bandwidth was kept between min/max. @fibann can you remember maybe?

Note also that there are several subtleties around bitrate:
- These are bitrates for RTP data sent by the peer connection according to the comment in Google's code, so would exclude non-RTP data.
- Literally every single WebRTC release since M71 has fixes related to bitrate, so there might be some bugs.
- There is probably a hard-coded limit for the minimum, although I would expect it to be lower than 2 Mbps (more like hundreds of kbps).

In short this needs investigation and I cannot confirm the behavior you're experiencing is expected or not. I still tend to think you should be able to go below 2 Mbps, since the default start bitrate is something like 800kpbs.

**Comment by astaikos316** [comment]
I did manage to trace what was going on and I am able to get down to 128kbps.  Related to this, I am trying to figure out what event or timing I can use to trigger sending data packets automatically after a peer to peer connection is made. I tried triggering on receiving an answer but it is too early and I get errors.

**Comment by djee-ms** [comment]
@astaikos316 can you please explain what was going on and how you went down to 128kpbs? Is there something we need to fix, or was it an error on your side? Why was it not working the first time?

On the data channel being open for writing, this depends I think whether the data channel is in-band (negotiated via an SDP session) or out-of-band (implicitly open with same ID on both peers). For the latter obviously there is no synchronization mechanism built in WebRTC since this is intrinsically an externally-synchronized case that the user has to deal with. For the former you can handle the `DataChannel.StateChanged` event and check that `DataChannel.State == ChannelState.Open` to confirm whether the data channel is ready for sending data.

**Comment by astaikos316** [comment]
@djee-ms I had an issue with one of my filter settings on my packet capture utility.  Once I spotted the issue I was able to see that I can get the bitrate to where i was expecting. 
Also, thank you for clarifying how I can check is the data channel is ready to send data.

**Comment by djee-ms** [comment]
Ok thanks for the feedback, glad it all works now! Closing as invalid since the error was on your side then. Please feel free to reopen if there's anything else.

**Comment by FakeHermes** [comment]
Hi @astaikos316 ,I want to know did you use "SetBitRate" at unity? if so, can you tell me how to use it?I did not found the setBitRate function at libs/unity/library/runtime/peerconntection.cs
thanks a lot~
